shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/ShardingSphereOptimizer.java

```diff
@@ -49,16 +61,27 @@ public RelNode optimize(final String sql) throws SQLParsingException {
             ShardingSphereSQLParserEngine sqlParserEngine = new ShardingSphereSQLParserEngine(DatabaseTypeRegistry.getTrunkDatabaseTypeName(context.getDatabaseType()));
             SqlNode sqlNode = SqlNodeConvertEngine.convert(sqlParserEngine.parse(sql, true));
             SqlNode validNode = context.getValidator().validate(sqlNode);
+            RelDataType resultType = context.getValidator().getValidatedNodeType(sqlNode);
             RelNode logicPlan = context.getRelConverter().convertQuery(validNode, false, true).rel;
-            return optimize(logicPlan);
+            return optimize(logicPlan, resultType);
         } catch (final UnsupportedOperationException ex) {
             throw new ShardingSphereException(ex);
         }
     }
     
-    private RelNode optimize(final RelNode logicPlan) {
+    private RelNode optimize(final RelNode logicPlan, final RelDataType resultType) {
         RelOptPlanner planner = context.getRelConverter().getCluster().getPlanner();
-        planner.setRoot(planner.changeTraits(logicPlan, context.getRelConverter().getCluster().traitSet().replace(EnumerableConvention.INSTANCE)));
-        return planner.findBestExp();
+        RelNode node = planner.changeTraits(logicPlan, context.getRelConverter().getCluster().traitSet().replace(EnumerableConvention.INSTANCE));
+        RelRoot root = constructRoot(node, resultType);
+        FederateOptimizer optimizer = new FederateOptimizer(planner);
+        RelRoot result = optimizer.optimize(root);
+        return result.rel;
+    }
+
+    private RelRoot constructRoot(final RelNode node, final RelDataType resultType) {
+        final RelDataType rowType = node.getRowType();
+        final List<Pair<Integer, String>> fields = Pair.zip(ImmutableIntList.identity(rowType.getFieldCount()), rowType.getFieldNames());
```

Comment:
@wgy8283335 Can we remove this useless final modifier?

Reply:
ok.